cluster/swarm/cluster.go

```diff
@@ -359,9 +359,13 @@ func (c *Cluster) removeEngine(addr string) bool {
 func (c *Cluster) monitorDiscovery(ch <-chan discovery.Entries, errCh <-chan error) {
 	// Watch changes on the discovery channel.
 	currentEntries := discovery.Entries{}
-	for {
+	for ch != nil || errCh != nil {
 		select {
-		case entries := <-ch:
+		case entries, ok := <-ch:
+			if !ok {
+				ch = nil
```

Comment:
@dongluochen 
> You can just return when any channel is closed.

Maybe it is appropriate for this case, but personally I don't think it is reasonable. If `select` is polling on `5` channels, return the function when `1` channel is closed while others are stilll available doesn't make sense.

Thanks very much!